File: SOLID/LSP/MainDocEditor.java
class Document {

    protected boolean isOpen = false;

    public void open() {
        isOpen = true;
        System.out.println("Document opened");
    }

    public void edit() {
        if (!isOpen) {
            throw new IllegalStateException("Document must be open");
        }
        System.out.println("Editing document");
    }

    public void close() {
        isOpen = false;
        System.out.println("Document closed");
    }
}
class ReadOnceDocument extends Document {

    private boolean alreadyOpened = false;

    @Override
    public void open() {
        if (alreadyOpened) {
            throw new IllegalStateException("Document cannot be reopened");
        }
        alreadyOpened = true;
        super.open();
    }
}


public class MainDocEditor{
    public static void main(String[] args){
        Document doc = new ReadOnceDocument();
        doc.open();
        doc.close();
        doc.open();  // runtime failure
    }
}

File: SOLID/LSP/python/MainDocEditor.py
from abc import ABC

class Document:
    def __init__(self):
        self._is_open = False

    def open(self):
        self._is_open = True
        print("Document opened")

    def edit(self):
        if not self._is_open:
            raise RuntimeError("Document must be open")
        print("Editing document")

    def close(self):
        self._is_open = False
        print("Document closed")


class ReadOnceDocument(Document):
    def __init__(self):
        super().__init__()
        self._already_opened = False

    def open(self):
        if self._already_opened:
            raise RuntimeError("Document cannot be reopened")
        self._already_opened = True
        super().open()


# Main
doc: Document = ReadOnceDocument()
doc.open()
doc.close()
doc.open()  # ❌ runtime error
